ESLint: Simplify via configuration cascading

jsnmoon opened the pull request:
### Changes proposed in this Pull Request:
Simplify ESLint configuration by cascading our configuration files. This significantly simplifies the `yarn lint` command.

See https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?

No, this does not add/remove features from Jetpack.

#### Testing instructions:

Run `yarn lint` and ensure that it outputs the same results as the current `master` branch.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->

No changelog necessary.

Commit:
ESLint: Simplify via configuration cascading

Simplify linting configuration and `yarn lint` by using config cascading.

See https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy.

Commit:
Simplify pre-commit hook script

jeherve reviewed (dismissed):
This seems to work well in my tests. 👍

simison commented on package.json:
```diff
@@ -52,9 +52,7 @@
-		"lint": "yarn lint-es6 && yarn lint-es5",
-		"lint-es5": "eslint _inc/*.js modules -c modules/.eslintrc.js",
-		"lint-es6": "eslint --ext .js --ext .jsx ./*.js _inc/client extensions -c .eslintrc.js",
+		"lint": "eslint --ext .js,.jsx ./*.js _inc/*.js _inc/client extensions modules",
```
Curious why `./*.js _inc/*.js` here if we already have `--ext`?

Should this be something like `. _inc extensions modules`?

simison replied:
Suppose root level folder covers sub-folders? i.e. no need `_inc/*/**`

brbrr replied:
I think `--ext` means the list of file types to lint, while `./*.js _inc/*.js` are wildcard file paths.

simison replied:
Docs suggest that both folders and globs work: https://eslint.org/docs/user-guide/command-line-interface (tho seems to be recommended to have them in quotes to ensure they're the same across environments).

At any rate, no action needed in this PR. :-)

jsnmoon replied:
@simison: `. _inc extensions modules` seems much better for its simplicity. I'll go ahead and make the change.

simison commented:
Personally, I find `overrides` section over individual files for _much_ better readability and grokking what actually is going on with linting in the repository.

That said, I appreciate that it might be impossible to organize linting that way in Jetpack because of such crazy variety of linting we need to do per-folder basis. :-/

simison commented on extensions/.eslintrc.js:
```diff
@@ -0,0 +1,26 @@
+module.exports = {
```
Please remember to sync this file to wpcom in D31208-code

simison replied:
>Please remember to sync this file to wpcom in D31208-code

Ahh nope, I forgot that we do linting there little differently. :-)

Commit:
Temporarily add Instant Search to .eslintignore

Commit:
Simplify lint command

Commit:
Fix newly discovered linting violations

jsnmoon commented:
I changed `yarn lint` to include more files and fixed previously unnoticed linting violations in `_inc/lib/tracks/`. 

This is ready for another round of reviews!

jeherve reviewed (approved):
This still works for me. 👍

simison commented:
Just wanted to note that `yarn lint` now brings 100+ lint warnings — would be nice to get those sorted out asap. They seem pretty trivial luckily. :-)

jsnmoon commented:
Thanks for the quick 👀 and 🚢, everyone!
